pkg/controller/keycloakbackup/keycloakbackup_controller.go

```diff
@@ -0,0 +1,170 @@
+		IsController: true,
+		OwnerType:    &kc.KeycloakBackup{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKeycloakBackup implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKeycloakBackup{}
+
+// ReconcileKeycloakBackup reconciles a KeycloakBackup object
+type ReconcileKeycloakBackup struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client   client.Client
+	scheme   *runtime.Scheme
+	context  context.Context
+	cancel   context.CancelFunc
+	recorder record.EventRecorder
+}
+
+// Reconcile reads that state of the cluster for a KeycloakBackup object and makes changes based on the state read
+func (r *ReconcileKeycloakBackup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeycloakBackup")
+
+	// Fetch the KeycloakBackup instance
+	instance := &kc.KeycloakBackup{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	currentState := common.NewBackupState()
+	currentState.Read(r.context, instance, r.client)
+
+	reconciler := NewKeycloakBackupReconciler()
+	desiredState := reconciler.Reconcile(currentState, instance)
+
+	actionRunner := common.NewClusterActionRunner(r.context, r.client, r.scheme, instance)
+	err = actionRunner.RunAll(desiredState)
+	if err != nil {
+		return r.ManageError(instance, err)
+	}
+
+	return r.ManageSuccess(instance, currentState)
+}
+
+func (r *ReconcileKeycloakBackup) ManageError(instance *kc.KeycloakBackup, issue error) (reconcile.Result, error) {
+	r.recorder.Event(instance, "Warning", "ProcessingError", issue.Error())
+
+	instance.Status.Message = issue.Error()
+	instance.Status.Ready = false
+	instance.Status.Phase = kc.BackupPhaseFailing
+
+	err := r.client.Status().Update(r.context, instance)
+	if err != nil {
+		log.Error(err, "unable to update status")
+	}
+
+	return reconcile.Result{
+		RequeueAfter: RequeueDelayErrorSeconds,
+		Requeue:      true,
+	}, nil
+}
+
+func (r *ReconcileKeycloakBackup) ManageSuccess(instance *kc.KeycloakBackup, currentState *common.BackupState) (reconcile.Result, error) {
+	isReady := true
+	if currentState.LocalPersistentVolumeJob != nil && currentState.LocalPersistentVolumeJob.Status.Succeeded != 1 {
+		isReady = false
+	}
+	if currentState.AwsJob != nil && currentState.AwsJob.Status.Succeeded != 1 {
+		isReady = false
+	}
```

Comment:
Would be good to abstract the success for each resource type. I have done this for some in the cluster_state.go. Could do the same there for the backupState

Reply:
Fixed.